### ISSUE TITLE
What is the advantage of FlashAttention-2 compared to PyTorch-2?

### ISSUE THREAD

**Issue by jun-fang** [Dao-AILab/flash-attention#462]
Hello, are the PyTorch results in your FlashAttention-2 paper based on PyTorch-2 with `torch.compile`? If not, what is the advantage of FlashAttention-2 compared to PyTorch-2?

I recently tested the GPU memory and training throughput for ViTs (base, large, and huge with input size 224) on 8x A100, PyTorch-2.0.1, CUDA-11.8 by using 
(1) **FA2 + PT2-compile**: FlashAttention-2 on top of PyTorch-2 with `torch.compile`, and 
(2) **PT2-compile**: PyTorch-2 with `torch.compile` only, no FlashAttention-2

Here are some of my observations: 
1.  **FA2 + PT2-compile** can save about 20% of GPU memory when compared to **PT2-compile**
2. Under the same batch size, or increasing the batch size by around 1.3x with the saved 20% GPU memory, **FA2 + PT2-compile** did not achieve training throughput speedups when compared to **PT2-compile**. Actually, **FA2 + PT2-compile** is even slower than **PT2-compile** sometimes

Are these above results the expected for ViTs with the seqlen of about ~200 (196+1 for ViT-base and 256+1 for ViT-huge)? If not, can you share more speedup results for ViTs and the correct ways to utilize FlashAttention-2 on top of PyTorch-2? 

Thank you in advance!

**Comment by tridao** [comment]
ViT has very small seqlen (200) so other things matter more while attention doesn't take that much time. You can take a look at how we implement [ViT](https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/models/vit.py) (it's been a while since I touched this file so I'm not 100% sure it still works). Last time I tested this (maybe 6 months ago) it was 1.5x faster than the implementation from timm, but of course things have changed with pytorch version and timm version etc.

Things like fused residual and layernorm, or fusing matmul and bias might matter more here. Presumable torch compile does some of this, I'm not sure.

**Comment by tridao** [comment]
I just fixed sth minor so that ViT runs, you can see the test [here](https://github.com/Dao-AILab/flash-attention/blob/main/tests/models/test_vit.py). I did a quick benchmark on A100 80GB PCIe and our implementation is about 1.35x faster than timm (without torch.compile), on both the forward and backward.
I couldn't benchmark torch.compile because it's hanging for some reason. If you get to benchmark it lmk.

Back to your question, maybe using FA2 in your model is causing torch.compile to insert some graph breaks that interferes with what's torch.compile is doing. I'm just speculating, I don't know torch.compile very well.

**Comment by zaptrem** [comment]
@tridao It seems that flash-attention's code style does indeed force graph breaks into an otherwise unbroken model. When full graph is enabled the following is thrown:

```
torch._dynamo.exc.Unsupported: call_method UserDefinedObjectVariable(fwd) __call__ [TensorVariable(), TensorVariable(), TensorVariable(), ConstantVariable(NoneType), ConstantVariable(float), ConstantVariable(float), ConstantVariable(bool), ConstantVariable(int), ConstantVariable(int), ConstantVariable(bool), ConstantVariable(NoneType)] {}

from user code:
   File "/mnt/wsl/PhysicalDrive1p1/dance5/x-transformers/x_transformers/continuous.py", line 127, in forward
    x, intermediates = self.attn_layers(x, mask = mask, mems = mems, return_hiddens = True, **kwargs)
  File "/mnt/wsl/PhysicalDrive1p1/dance5/the-model/env/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/mnt/wsl/PhysicalDrive1p1/dance5/x-transformers/x_transformers/x_transformers.py", line 1325, in forward
    block_return = block(x, mask = mask, context_mask = self_attn_kv_mask, attn_mask = attn_mask, rel_pos = self.rel_pos, rotary_pos_emb = rotary_pos_emb, prev_attn = prev_attn, cache = None, mem = layer_mem, return_intermediates = True) # cache = next(iter_attn_cache, None)
  File "/mnt/wsl/PhysicalDrive1p1/dance5/the-model/env/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/mnt/wsl/PhysicalDrive1p1/dance5/x-transformers/x_transformers/x_transformers.py", line 936, in forward
    out, intermediates = self.attend(
  File "/mnt/wsl/PhysicalDrive1p1/dance5/the-model/env/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/mnt/wsl/PhysicalDrive1p1/dance5/x-transformers/x_transformers/attend.py", line 294, in forward
    return self.flash_attn(q, k, v, mask = mask, attn_bias = attn_bias)
  File "/mnt/wsl/PhysicalDrive1p1/dance5/x-transformers/x_transformers/attend.py", line 230, in flash_attn
    out = flash_attn_func(q, k, v, dropout_p=0.0, causal=causal) # window_size=window_size
  File "/mnt/wsl/PhysicalDrive1p1/dance5/the-model/env/lib/python3.11/site-packages/flash_attn/flash_attn_interface.py", line 705, in flash_attn_func
    return FlashAttnFunc.apply(
  File "/mnt/wsl/PhysicalDrive1p1/dance5/the-model/env/lib/python3.11/site-packages/flash_attn/flash_attn_interface.py", line 434, in forward
    out, q, k, v, out_padded, softmax_lse, S_dmask, rng_state = _flash_attn_forward(
  File "/mnt/wsl/PhysicalDrive1p1/dance5/the-model/env/lib/python3.11/site-packages/flash_attn/flash_attn_interface.py", line 47, in _flash_attn_forward
    out, q, k, v, out_padded, softmax_lse, S_dmask, rng_state = flash_attn_cuda.fwd(

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
```

**Comment by tridao** [comment]
Do you know how to make things work with torch.compile?

**Comment by zaptrem** [comment]
> Do you know how to make things work with torch.compile?

Only what I've discovered via trial and error over the past few days. The above issue is remedied by not using that unwrapping syntax (e.g., instead of `thing, other_thing = tuple_to_unwrap` do `thing = tuple_to_unwrap[0]` `other_thing = tuple_to_unwrap[1]`)

[Some issues](https://github.com/pytorch/pytorch/issues/116935) deeper in Inductor are currently over my head, though.
